src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_18_0.png)
+    
+
+
+For ROC we see no improvement, whereas higher number of trees works better than just a few. This model didn't improve our baseline defined by the linear model in blue.
+
+In Precision-Recall, we obtain similar AUC for 200 trees, with better results for RandomForest between 0-15% recall, and performing worse than logistic regression beyond 20% recall (we see blue and brown lines intersecting at that point). This would be interesting if we knew in advance the exact point we would like to be in, which we know more or less but will need to make more adjustments. 
+
+As a result, no model has yet defeated our beatiful logistic regression. It is also worth to mention that training for this n_trees took around 20min, while lr run x10 faster - in case scalability and deployment plays an important role.
+
+
+```python
+def plot_feature_importance(model, feature_names, title="Feature Importance"):
+    """
+    Plot the feature importances of a tree-based model.
+
+    Parameters:
+    - model: Trained tree-based model (like RandomForest or DecisionTree).
+    - feature_names: List of feature names.
+    - title: Title of the plot.
+    """
+    importances = model.feature_importances_
+    indices = np.argsort(importances)[::-1]
+
+    plt.figure(figsize=(10, 6))
+    plt.title(title)
+    plt.bar(range(len(feature_names)), importances[indices], align="center")
+    plt.xticks(range(len(feature_names)), np.array(feature_names)[indices], rotation=90)
+    plt.xlim([-1, len(feature_names)])
+    plt.show()
+
+```
+
+
+```python
+X_train, y_train, X_val, y_val, X_test, y_test = temporal_data_split(data_processed, cols=cols)
+feature_names = X_train.columns
+```
+
+    INFO - Date from: 2020-10-05
+    INFO - Date to: 2021-03-03
+    INFO - Train set ratio: 73.44%
+    INFO - Validation set ratio: 13.32%
+    INFO - Test set ratio: 13.24%
+
+
+
+```python
+# Train a RandomForestClassifier with 200 estimators
+rf = RandomForestClassifier(n_estimators=200)
+rf.fit(X_train, y_train)
+
+# Assuming you have already defined the feature_names variable
+# If not, you can define it like this:
+# feature_names = X_train.columns
+
+# Plot feature importance for the trained RandomForest model
+plot_feature_importance(rf, feature_names, title="Random Forest (n_trees=200) Feature Importance")
+
+# Display the feature importance plot
+plt.show()
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_22_0.png)
+    
+
+
+
+If performance could be a significant issue in this case, we can consider discarding from 'count_pets' to the right at a later time. However, we would continue in this case with all the cols
+
+## Gradient Boosting Trees
+
+
+```python
+from sklearn.ensemble import GradientBoostingClassifier
+
+# Let's be conservative for reducing total time
+n_trees_values = [5, 100, 250]
+max_depth_values = [1, 3, 5]
```

Comment:
I'd say depth 1 is too small

Reply:
Also bear in mind that the most important parameter of GBM is the learning rate (same for NN), so feel free to play with it